File: widgets/top_nav_bar.py
"""
顶部导航栏组件。
显示应用标题、相机连接状态、采集进度条、和设置/帮助/连接相机按钮。
"""

from PyQt5.QtCore import Qt, pyqtSignal
from PyQt5.QtWidgets import (
    QWidget, QHBoxLayout, QLabel, QPushButton, QProgressBar, QSizePolicy,
)

import resources as res


class TopNavBar(QWidget):
    """顶部导航栏，包含标题、进度、相机状态和功能按钮"""

    # 信号定义
    settings_clicked = pyqtSignal()          # 设置按钮被点击
    help_clicked = pyqtSignal()              # 帮助按钮被点击
    connect_camera_clicked = pyqtSignal()    # 连接相机按钮被点击（断开状态时）
    disconnect_camera_clicked = pyqtSignal() # 断开相机按钮被点击（连接状态时）

    def __init__(self, parent=None):
        super().__init__(parent)
        # 固定高度 48px，宽度自适应
        self.setFixedHeight(48)
        self.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Fixed)

        # 顶栏背景色 + 底部分割线
        self.setStyleSheet(f"""
            TopNavBar {{
                background-color: {res.BG_MAIN};
                border-bottom: 1px solid {res.BORDER_DEFAULT};
            }}
        """)

        # 水平布局：左侧标题 | 中部进度 | 右侧按钮
        layout = QHBoxLayout(self)
        layout.setContentsMargins(24, 0, 24, 0)
        layout.setSpacing(12)

        # ── 左侧：机器人图标 + 应用标题 ──
        # 使用 Unicode 字符作为图标占位（避免外部图片依赖）
        icon_label = QLabel("\U0001f916")  # 机器人 emoji
        icon_label.setStyleSheet(f"font-size: 20px; color: {res.PRIMARY}; border: none;")
        layout.addWidget(icon_label)

        # 应用标题：H1 字号，深色粗体
        title = QLabel("手眼标定数据收集助手 V1.0")
        title.setStyleSheet(f"""
            font-size: {res.FONT_H1}px;
            font-weight: 600;
            color: {res.TEXT_TITLE};
            border: none;
        """)
        layout.addWidget(title)

        layout.addStretch()  # 弹性空间，将后续组件推到中间和右侧

        # ── 中部：相机连接状态 ──
        self._camera_status_label = QLabel("⚪ 未连接")  # ⚪ 未连接
        self._camera_status_label.setStyleSheet(
            f"font-size: {res.FONT_HINT}px; color: {res.TEXT_HINT}; border: none;"
        )
        layout.addWidget(self._camera_status_label)

        # 连接相机按钮
        self._connected = False  # 跟踪当前连接状态
        self._btn_connect = QPushButton("连接相机")  # 连接相机
        self._btn_connect.setFixedHeight(28)
        self._btn_connect.setStyleSheet(f"""
            QPushButton {{
                background-color: {res.PRIMARY};
                color: #FFFFFF;
                border: none;
                border-radius: 4px;
                font-size: {res.FONT_HINT}px;
                padding: 0 12px;
            }}
            QPushButton:hover {{
                background-color: {res.PRIMARY_HOVER};
            }}
        """)
        # 根据当前连接状态分发不同的信号
        self._btn_connect.clicked.connect(self._on_connect_btn_clicked)
        layout.addWidget(self._btn_connect)

        layout.addSpacing(24)

        # ── 中部：采集进度标签 + 进度条 ──
        self._progress_label = QLabel("已采集 0 组数据")  # 已采集 0 组数据
        self._progress_label.setStyleSheet(
            f"font-size: {res.FONT_BODY}px; color: {res.TEXT_BODY}; border: none;"
        )
        layout.addWidget(self._progress_label)

        # 进度条：天蓝色填充，6px 高度，固定宽度 120px
        self._progress_bar = QProgressBar()
        self._progress_bar.setFixedWidth(120)
        self._progress_bar.setFixedHeight(6)
        self._progress_bar.setRange(0, 100)
        self._progress_bar.setValue(0)
        self._progress_bar.setTextVisible(False)
        layout.addWidget(self._progress_bar)

        layout.addStretch()

        # ── 右侧：设置按钮 + 帮助按钮 ──
        for text, signal, tip in [
            ("⚙", self.settings_clicked, "设置"),   # ⚙ 设置
            ("?", self.help_clicked, "帮助"),            # ? 帮助
        ]:
            btn = QPushButton(text)
            btn.setFixedSize(24, 24)
            btn.setToolTip(tip)
            btn.setStyleSheet(f"""
                QPushButton {{
                    border: none;
                    font-size: 18px;
                    color: {res.TEXT_HINT};
                }}
                QPushButton:hover {{
                    color: {res.PRIMARY};
                }}
            """)
            btn.clicked.connect(signal.emit)
            layout.addWidget(btn)

    def update_progress(self, current, total=None):
        """
        更新采集进度显示。
        current: 当前已采集组数
        total: 目标组数（None 时取 current 和 15 的较大值）
        """
        if total is None or total <= 0:
            total = max(current, res.RECOMMENDED_COUNT)
        self._progress_label.setText(f"已采集 {current}/{total} 组数据")
        # 计算百分比进度
        pct = min(int(current / max(total, 1) * 100), 100)
        self._progress_bar.setValue(pct)

    def set_camera_status(self, connected, device_name=""):
        """
        更新相机连接状态显示。
        connected=True 时显示绿色圆点 + 设备名，按钮变为"断开相机"；
        connected=False 时显示灰色圆点 + "未连接"，按钮变为"连接相机"。
        """
        self._connected = connected
        if connected:
            self._camera_status_label.setText(
                f"\U0001f7e2 {device_name}"  # 🟢 设备名
            )
            self._camera_status_label.setStyleSheet(
                f"font-size: {res.FONT_HINT}px; color: {res.SUCCESS}; border: none;"
            )
            self._btn_connect.setText("断开相机")  # 断开相机
            self._btn_connect.setStyleSheet(f"""
                QPushButton {{
                    background-color: {res.BG_MAIN};
                    color: {res.ERROR};
                    border: 1px solid {res.ERROR};
                    border-radius: 4px;
                    font-size: {res.FONT_HINT}px;
                    padding: 0 12px;
                }}
                QPushButton:hover {{
                    background-color: {res.ERROR_BG};
                }}
            """)
        else:
            self._camera_status_label.setText("⚪ 未连接")  # ⚪ 未连接
            self._camera_status_label.setStyleSheet(
                f"font-size: {res.FONT_HINT}px; color: {res.TEXT_HINT}; border: none;"
            )
            self._btn_connect.setText("连接相机")  # 连接相机
            self._btn_connect.setStyleSheet(f"""
                QPushButton {{
                    background-color: {res.PRIMARY};
                    color: #FFFFFF;
                    border: none;
                    border-radius: 4px;
                    font-size: {res.FONT_HINT}px;
                    padding: 0 12px;
                }}
                QPushButton:hover {{
                    background-color: {res.PRIMARY_HOVER};
                }}
            """)

    def _on_connect_btn_clicked(self):
        """
        连接/断开按钮点击处理。
        根据当前连接状态分发不同的信号。
        """
        if self._connected:
            self.disconnect_camera_clicked.emit()
        else:
            self.connect_camera_clicked.emit()

File: src/ui/TopNavBar.cpp
#include "ui/TopNavBar.h"
#include "ui/Theme.h"
#include <QHBoxLayout>

TopNavBar::TopNavBar(QWidget* parent)
    : QWidget(parent)
{
    setFixedHeight(48);
    setStyleSheet(QStringLiteral("background-color: %1; border-bottom: 1px solid %2;")
                  .arg(Theme::BG_MAIN).arg(Theme::BORDER_DEFAULT));

    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(24, 0, 24, 0);
    layout->setSpacing(16);

    // Title
    m_title = new QLabel(QStringLiteral("手眼标定数据收集助手 V1.0"), this);
    m_title->setStyleSheet(QStringLiteral("font-size: %1px; font-weight: 600; color: %2; border: none;")
                           .arg(Theme::FONT_H1).arg(Theme::TEXT_TITLE));
    layout->addWidget(m_title);

    // Progress
    m_progressLabel = new QLabel(QStringLiteral("已采集 0/15 组数据"), this);
    m_progressLabel->setStyleSheet(QStringLiteral("font-size: %1px; color: %2; border: none;")
                                   .arg(Theme::FONT_BODY).arg(Theme::TEXT_BODY));
    layout->addWidget(m_progressLabel);

    m_progressBar = new QProgressBar(this);
    m_progressBar->setFixedSize(120, 6);
    m_progressBar->setRange(0, Theme::RECOMMENDED_COUNT);
    m_progressBar->setValue(0);
    m_progressBar->setTextVisible(false);
    layout->addWidget(m_progressBar);

    layout->addStretch();

    // Camera status
    m_cameraDot = new QLabel(this);
    m_cameraDot->setFixedSize(10, 10);
    m_cameraDot->setStyleSheet(QStringLiteral("background-color: %1; border-radius: 5px;").arg(Theme::TEXT_HINT));
    layout->addWidget(m_cameraDot);

    m_cameraLabel = new QLabel(QStringLiteral("未连接"), this);
    m_cameraLabel->setStyleSheet(QStringLiteral("font-size: %1px; color: %2; border: none;")
                                 .arg(Theme::FONT_HINT).arg(Theme::TEXT_HINT));
    layout->addWidget(m_cameraLabel);

    m_btnConnect = new QPushButton(QStringLiteral("连接"), this);
    m_btnConnect->setFixedSize(60, 28);
    m_btnConnect->setStyleSheet(QStringLiteral(R"(
        QPushButton { color: %1; background: transparent; border: 1px solid %1;
                      border-radius: 4px; font-size: %2px; }
        QPushButton:hover { border-color: %3; color: %3; }
    )").arg(Theme::PRIMARY).arg(Theme::FONT_HINT).arg(Theme::PRIMARY_HOVER));
    layout->addWidget(m_btnConnect);

    // Settings / Help
    m_btnSettings = new QPushButton(QStringLiteral("设置"), this);
    m_btnSettings->setFixedSize(48, 28);
    m_btnSettings->setStyleSheet(QStringLiteral("color: %1; background: transparent; border: none; font-size: %2px;")
                                 .arg(Theme::TEXT_HINT).arg(Theme::FONT_HINT));
    m_btnHelp = new QPushButton(QStringLiteral("帮助"), this);
    m_btnHelp->setFixedSize(48, 28);
    m_btnHelp->setStyleSheet(QStringLiteral("color: %1; background: transparent; border: none; font-size: %2px;")
                             .arg(Theme::TEXT_HINT).arg(Theme::FONT_HINT));

    layout->addWidget(m_btnSettings);
    layout->addWidget(m_btnHelp);

    connect(m_btnConnect,  &QPushButton::clicked, this, &TopNavBar::onConnectBtnClicked);
    connect(m_btnSettings, &QPushButton::clicked, this, &TopNavBar::settingsClicked);
    connect(m_btnHelp,     &QPushButton::clicked, this, &TopNavBar::helpClicked);
}

void TopNavBar::updateProgress(int current, int total)
{
    int t = (total > 0) ? total : Theme::RECOMMENDED_COUNT;
    m_progressLabel->setText(QStringLiteral("已采集 %1/%2 组数据").arg(current).arg(t));
    m_progressBar->setMaximum(t);
    m_progressBar->setValue(current);
}

void TopNavBar::setCameraStatus(bool connected, const QString& deviceName)
{
    m_cameraConnected = connected;
    m_btnConnect->setEnabled(true);
    if (connected) {
        m_cameraDot->setStyleSheet(QStringLiteral("background-color: %1; border-radius: 5px;").arg(Theme::SUCCESS));
        m_cameraLabel->setText(deviceName.isEmpty() ? QStringLiteral("已连接") : deviceName);
        m_cameraLabel->setStyleSheet(QStringLiteral("font-size: %1px; color: %2; border: none;")
                                     .arg(Theme::FONT_HINT).arg(Theme::SUCCESS));
        m_btnConnect->setText(QStringLiteral("断开"));
    } else {
        m_cameraDot->setStyleSheet(QStringLiteral("background-color: %1; border-radius: 5px;").arg(Theme::TEXT_HINT));
        m_cameraLabel->setText(QStringLiteral("未连接"));
        m_cameraLabel->setStyleSheet(QStringLiteral("font-size: %1px; color: %2; border: none;")
                                     .arg(Theme::FONT_HINT).arg(Theme::TEXT_HINT));
        m_btnConnect->setText(QStringLiteral("连接"));
    }
}

void TopNavBar::setConnectBusy(bool busy, const QString& text)
{
    if (busy) {
        m_btnConnect->setText(text);
        m_btnConnect->setEnabled(false);
    } else {
        m_btnConnect->setEnabled(true);
        m_btnConnect->setText(m_cameraConnected ? QStringLiteral("断开") : QStringLiteral("连接"));
    }
}

void TopNavBar::onConnectBtnClicked()
{
    if (m_cameraConnected)
        emit disconnectCameraClicked();
    else
        emit connectCameraClicked();
}
